src/main/scala/firrtl/transforms/InlineBooleanExpressions.scala

```diff
@@ -0,0 +1,163 @@
+// See LICENSE for license details.
+
+package firrtl
+package transforms
+
+import firrtl.annotations.Target
+import firrtl.annotations.TargetToken.{fromStringToTargetToken, OfModule, Ref}
+import firrtl.ir._
+import firrtl.passes.{InferTypes, LowerTypes, SplitExpressions}
+import firrtl.options.Dependency
+import firrtl.PrimOps._
+import firrtl.WrappedExpression._
+
+import scala.collection.mutable
+
+/** Inline Bool expressions
+  *
+  * The following conditions must be satisfied to inline
+  * 1. has type [[Utils.BoolType]]
+  * 2. is bound to a [[DefNode]] with name starting with '_'
+  * 3. is bound to a [[DefNode]] with the same source locator
+  * 4. the resulting expression can be emitted to verilog without needing any
+  *    parentheses
+  */
+class InlineBooleanExpressions extends Transform with DependencyAPIMigration {
+
+  override def prerequisites = Seq(
+    Dependency(InferTypes),
+    Dependency(LowerTypes)
+  )
+
+  override def optionalPrerequisites = Seq(
+    Dependency(SplitExpressions)
+  )
+
+  override def optionalPrerequisiteOf = Seq.empty
+
+  override def invalidates(a: Transform) = a match {
+    case _: DeadCodeElimination => true
+    case _ => false
+  }
+
+  type Netlist = mutable.HashMap[(WrappedExpression, Info), Expression]
+
+  private def isArgN(outerExpr: DoPrim, subExpr: Expression, n: Int): Boolean = {
+    outerExpr.args.lift(n) match {
+      case Some(arg) => arg eq subExpr
+      case _ => false
+    }
+  }
+
+  /** Maps a [[PrimOp]] to a precedence number, lower number means higher precedence
+    *
+    * Only the [[PrimOp]]s contained in this map will be inlined. [[PrimOp]]s
+    * like [[PrimOp.Neg]] are not in this map because inlining them may result
+    * in illegal verilog like '--2sh1'
+    */
+  private val precedenceMap: Map[PrimOp, Int] = {
+    val precedenceSeq = Seq(
```

Comment:
Why not just emit parens in the Verilog?

Reply:
the main purpose of this change is to make the verilog pretty. and emitting perens would require more changes to the verilog emitter.

Reply:
It would make the Verilog more human-readable, though.

Reply:
I think its fine to wait on a future PR to decide when to add parens.

Reply:
My suggestion is "always use parens," since code that relies on understanding of operator precedence is hard to read, and the goal is nominally to produce readable Verilog. Also, the Emitter changes add a large waiver to a generally good check -- the emitter will now emit any expression tree of bool type without complaining, even though this may produce broken Verilog. Using parens for all expression nesting would address both of these.

Reply:
@albertchen-sifive can we see the result with parantheses, and then together make a decision?